thriftrw/spec/enum.pyx

```diff
@@ -153,13 +153,19 @@ cdef class EnumTypeSpec(TypeSpec):
         writer.write_i32(value)
 
     cpdef object from_primitive(self, object prim_value):
+        val = self.items.get(prim_value)
+        if val is not None:
+            return val
         return prim_value
 
     cpdef void validate(self, object instance) except *:
-        if instance not in self.values_to_names:
-            raise ValueError(
-                '%r is not a valid value for enum "%s"' % (instance, self.name)
-            )
+        if instance in self.values_to_names:
+            return
+        if instance in self.items:
+            return
```

Comment:
I don't know if this is right. validate decides if the object is of the right type so that it can be serialized/deserialized by ThriftRW. A struct with a string instead of an int for an enum field won't serialize because both, to_wire and write_to assume it's an integer. (Adding a check for that on the hot path is definitely not something we should do.) I think validate on strings should still be false because we don't pass enums around as strings.

Reply:
Hmm right - I only made this change because the original error I got was this, which I assumed meant that I needed to also change validate: 
```
Traceback (most recent call last):
  File "manager.py", line 321, in blocking_load_task_response_from_file
    return scheduler.Response.from_primitive(from_json)
  File "thriftrw/spec/common.pyx", line 74, in thriftrw.spec.common.from_primitive_classmethod.from_primitive (thriftrw/spec/common.c:2062)
  File "thriftrw/spec/struct.pyx", line 195, in thriftrw.spec.struct.StructTypeSpec.from_primitive (thriftrw/spec/struct.c:3862)
  File "thriftrw/spec/struct.pyx", line 202, in thriftrw.spec.struct.StructTypeSpec.from_primitive (thriftrw/spec/struct.c:3768)
  File "thriftrw/spec/list.pyx", line 76, in thriftrw.spec.list.ListTypeSpec.from_primitive (thriftrw/spec/list.c:2440)
  File "thriftrw/spec/list.pyx", line 77, in thriftrw.spec.list.ListTypeSpec.from_primitive (thriftrw/spec/list.c:2384)
  File "thriftrw/spec/struct.pyx", line 202, in thriftrw.spec.struct.StructTypeSpec.from_primitive (thriftrw/spec/struct.c:3768)
  File "thriftrw/spec/struct.pyx", line 195, in thriftrw.spec.struct.StructTypeSpec.from_primitive (thriftrw/spec/struct.c:3862)
  File "thriftrw/spec/struct.pyx", line 204, in thriftrw.spec.struct.StructTypeSpec.from_primitive (thriftrw/spec/struct.c:3807)
  File "thriftrw/spec/struct.pyx", line 359, in thriftrw.spec.struct.struct_init.__init__ (thriftrw/spec/struct.c:6247)
  File "thriftrw/spec/struct.pyx", line 206, in thriftrw.spec.struct.StructTypeSpec.validate (thriftrw/spec/struct.c:4325)
  File "thriftrw/spec/struct.pyx", line 222, in thriftrw.spec.struct.StructTypeSpec.validate (thriftrw/spec/struct.c:4244)
ValueError: Field 2 of Task is invalid: SOMESTRING' is not a valid value for enum "Enum"
```

Reply:
Also, as far as I can see, `from_wire` already has a type check so it shouldn't accept a string

Reply:
`validate` is connected to both, `from_wire` and `to_wire` (and the streaming
variants, `read_from` and `write_to`). The expectation is that if `validate()`
doesn't fail, I can serialize the value with the TypeSpec.

If you look at the `to_wire`/`write_to` implementations, you'll note that
they'll fail with strings.

`validate()` should check if this is the way we pass the value around. In
Python, regardless of primitive representation, the way enums are passed
around in memory is as integers, so that's all validate should allow.

Reply:
To elaborate, I believe this code will let me do,

    MyStruct(someEnum="FOO")

That will not fail validation but then when I try to actually serialize it
using ThriftRW's binary encoding, it will explode.

Reply:
(Only `MyStruct(someEnum=MyEnum.FOO)`, which is equivalent to `MyStruct(someEnum=42)` must be allowed.)

Reply:
Makes sense - I reverted the validate change and things seem to work as expected